openmdao/solvers/ln_direct.py

```diff
@@ -42,23 +91,123 @@ def solve(self, rhs_mat, system, mode):
-            ident = np.eye(n_edge)
+        sol_buf = OrderedDict()
 
-            partials = np.empty((n_edge, n_edge))
+        # TODO: This solver could probably work with multiple RHS
+        for voi, rhs in rhs_mat.items():
+            self.voi = None
 
-            for i in range(n_edge):
-                partials[:, i] = self.mult(ident[:, i])
+            # OpenMDAO does matrix vector product.
+            if self.options['jacobian_method'] == 'MVP':
+
+                self.mode = mode
+
+                n_edge = len(rhs)
+                ident = np.eye(n_edge)
+
+                partials = np.empty((n_edge, n_edge))
+
+                for i in range(n_edge):
+                    partials[:, i] = self.mult(ident[:, i])
+
+            # Assemble the Jacobian
+            else:
+
+                # Must clear the jacobian if we switch modes
+                if self.mode != mode:
+                    self.setup(system)
+                self.mode = mode
+                sys_name = system.name + '.'
+                conn = system.connections
+
+                partials = self.jacobian
+                u_vec = system.unknowns
+                var_names = u_vec.keys()
+
+                for sub in system.components(recurse=True, include_self=False):
+
+                    jac = sub._jacobian_cache
+
+                    # This method won't work on components where apply_linear
+                    # is overridden.
+                    if jac is None:
+                        msg = "The 'assemble' jacobian_method is not supported when " + \
+                             "'apply_linear' is used on a component (%s)." % sub.pathname
+                        raise RuntimeError(msg)
+
+                    sub_u = sub.unknowns
+                    sub_name = sub.pathname
+                    icache = self.icache
+
+                    for key in jac:
+                        o_var, i_var = key
+                        key2 = (sub_name, key)
+
+                        # Derivs wrt states, but states live in u_vec
+                        if i_var in sub.states:
+                            sub_p = sub_u
+                        else:
+                            sub_p = sub.params
+
+                        # Skip anything not relevant
+                        if o_var not in sub_u or i_var not in sub_p:
+                            continue
+
+                        # We cache the location of each variable in our jacobian
+                        if key2 not in icache:
+                            o_var_pro = sub_u.metadata(o_var)['top_promoted_name']
+                            meta = sub_p.metadata(i_var)
+                            i_var_pro = meta['top_promoted_name']
+
+                            # States are fine ...
+                            if i_var in sub.states:
+                                pass
+
+                            #... but inputs need to find their source.
+                            elif i_var_pro not in var_names:
```

Comment:
shouldn't do 'in var_names' here, since var_names is a list.  Just do 'in u_vec'.  The var_names temporary isn't needed at all.